File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    count_map = {}
    unique_set = set()

    for i in topping:
        count_map[i] = count_map.get(i, 0) + 1

    for i in topping:
        count_map[i] -= 1
        unique_set.add(i)
        if count_map[i] == 0:
            del count_map[i]
        if len(count_map) == len(unique_set):
            answer += 1

    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;
class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        Map<Integer, Integer> countMap = new HashMap<>();
        Set<Integer> set = new HashSet<>();

        for (int i : topping) {
            countMap.put(i, countMap.getOrDefault(i, 0) + 1);
        }

        for (int i : topping) {
            countMap.put(i, countMap.get(i) - 1);
            set.add(i);
            if (countMap.get(i) == 0) {
                countMap.remove(i);
            }
            if (countMap.size() == set.size()) {
                answer++;
            }
        }

        return answer;
    }
}